File: apps/api/app/schemas/events.py
"""Run-event contract. Shared by worker (producer), API SSE relay, and the web UI.

The TypeScript mirror lives in packages/shared/src/events.ts and is kept honest by
tests/integration/test_event_schema_compat.py.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

EventType = Literal[
    "run.started",
    "run.finished",
    "run.failed",
    "run.cancelled",
    "step.started",
    "step.progress",
    "step.finished",
    "tool.called",
    "tool.result",
    "file.changed",
    "test.run",
    "security.finding",
    "review.completed",
    "budget.update",
    "awaiting_approval",
]


class RunEvent(BaseModel):
    id: str = Field(description="monotonic per-run id; usable as SSE Last-Event-ID")
    run_id: uuid.UUID
    task_id: uuid.UUID
    type: EventType
    at: datetime
    # step context (present for step.* / tool.* / file.* events)
    agent: str | None = None
    step_seq: int | None = None
    # free-form, event-type-specific, already redacted of secrets by the producer
    data: dict[str, Any] = {}

File: packages/shared/src/events.ts
// Wire contract for run events. MUST stay in sync with
// apps/api/app/schemas/events.py (RunEvent). tests/integration/test_event_schema_compat.py
// asserts the field names match.

export type RunEventType =
  | "run.started"
  | "run.finished"
  | "run.failed"
  | "run.cancelled"
  | "step.started"
  | "step.progress"
  | "step.finished"
  | "tool.called"
  | "tool.result"
  | "file.changed"
  | "test.run"
  | "security.finding"
  | "review.completed"
  | "budget.update"
  | "awaiting_approval";

export interface RunEvent {
  id: string;
  run_id: string;
  task_id: string;
  type: RunEventType;
  at: string;
  agent: string | null;
  step_seq: number | null;
  data: Record<string, unknown>;
}

export const RUN_EVENT_FIELDS = [
  "id",
  "run_id",
  "task_id",
  "type",
  "at",
  "agent",
  "step_seq",
  "data",
] as const;
